### PR TITLE
Post Deletion API Route

### LINKED ISSUE #91: [Post view] [BE] Post deletion API route
## Context

We want to allow admins to delete posts. This can happen if the animal has found a permanent home, in which case the admins would like to clean up the feed by deleting posts they no longer want to keep on the platform.

## Requirements

Build an API endpoint at `post.delete` that takes in an input of:

- postOid: (use string or a zodOidType - the string length is fixed to 24 characters)

The API should do the following:

1. Fetch the db to get the corresponding post document using the _id
2. Use the attachments array to make a storage bucket call to delete the attachments associated with the post from our storage bucket (use the S3 API - you can try looking at `getDirectUploadUrl` for ideas on how to properly do this)
3. If the deletion was successful, then delete the post document using the _id (writing the corresponding database action might be necessary)

Write an integration test to test your implementation of this API endpoint.

#### PM

- [ ] The post should be deleted from the DB upon calling this API
- [ ] All associated data with an ID should be deleted (photos, videos, post details, etc.)
- [ ] A clear error message will be interpretable if the delete fails for whatever reason

#### EM

- [ ] Errors thrown where necessary
- [ ] Use bulk-deletion if possible
- [ ] Integration tests (actually hitting the storage bucket) written

### PR DESCRIPTION
**Pull request by uma-anand**
## Post Deletion API Route

Issue Number(s): #91

What does this PR change and why?
We want to allow admins to delete posts. This can happen if the animal has found a permanent home, in which case the admins would like to clean up the feed by deleting posts they no longer want to keep on the platform.

### Checklist

- [ ] Requirements have been implemented
- [ ] Acceptance criteria are met
- [ ] Database schema docs have been updated or are not necessary
- [ ] Code follows design and style guidelines
- [ ] Commits follow guidelines (concise, squashed, etc)
- [ ] Relevant reviewers (Senior Dev/EM/Designers) have been assigned to this PR

### CODE REVIEW

File: web/db/actions/Post.ts [hunk 1]
```diff
@@ -82,6 +82,26 @@ async function getResizedUploadUrl(uuid: string): Promise<string> {
...
 }
 
+async function deleteAttachments(keysToDelete: string[]) {
+  //will make key splitting more efficient later
+  const objectsToDelete = keysToDelete.map((keyToDelete) => ({
+    Key:
+      keyToDelete.split("/")[keyToDelete.split("/").length - 2] +
+      "/" +
+      keyToDelete.split("/")[keyToDelete.split("/").length - 1],
+  }));
```
Comment (katsukixyz) [inline web/db/actions/Post.ts]: Can you walk me through this? What does `keyToDelete` look like and what does it need to transform into?

File: web/db/actions/Post.ts [hunk 2]
```diff
@@ -82,6 +82,26 @@ async function getResizedUploadUrl(uuid: string): Promise<string> {
...
+  const deleteObjectsCommand = new DeleteObjectsCommand({
+    Bucket: consts.storageBucket,
+    Delete: { Objects: objectsToDelete },
+  });
+  try {
+    const returned = await storageClient.send(deleteObjectsCommand);
+    return { success: true };
+  } catch (error) {
+    return { success: false };
+  }
```
Comment (katsukixyz) [inline web/db/actions/Post.ts]: I think it'd be useful to have some retry logic here eventually too - `DeleteObjectsCommand` has a `quiet` mode that you can use to only get the keys for which the deletion request failed. If there are any keys in there, we should keep retrying (up to 3 times)

File: web/package.json [hunk 3]
```diff
@@ -67,6 +67,7 @@
     "react-responsive-carousel": "^3.2.23",
     "react-select": "^5.7.0",
     "react-table": "^7.8.0",
+    "sentry": "^0.1.2",
```
Comment (katsukixyz) [inline web/package.json]: This probably doesn't need to be installed - could you try following https://docs.sentry.io/platforms/javascript/guides/nextjs/ for the installation?

File: web/db/actions/Post.ts (hunk 1 above)
Comment (uma-anand) [reply web/db/actions/Post.ts]: So keyToDelete will look like this `https://pub-27cb50e8f369479a8f312921bab07d36.r2.dev/65024a562a5311729da58f7b/schedule%20(4).png`- `https://bucketurl/postid/imagename`. I just wanted the `postid/imagename` part of it because that's the key in the bucket. I just used that code when I was just working on the function. I will go in later to make it more efficient.

File: web/db/actions/Post.ts (hunk 2 above)
Comment (uma-anand) [reply web/db/actions/Post.ts]: Okay, will keep that in mind when I implement the rest of the function. Thanks!

File: web/db/actions/Post.ts (hunk 1 above)
Comment (katsukixyz) [reply web/db/actions/Post.ts]: Ah, got it. In that case, you can modify the `getPost` method to take in another boolean parameter `publicAttachmentUrls`. Change the logic so that only when `publicAttachmentUrls` is true, the modification on `post.attachments` is done. The keys themselves are stored in the database, so you won't need the key splitting logic after this change.

Comment (katsukixyz) [reply web/db/actions/Post.ts]: Make sure to update all uses of `getPost` after you make this change!

File: web/db/actions/Post.ts (hunk 2 above)
Comment (uma-anand) [reply web/db/actions/Post.ts]: So I have a couple of questions related to this. If we have a couple of images that we delete associated with a post, but not all of them- how do we proceed? Do we continue with deleting the object or do we try to restore the pictures we deleted because we couldn't delete all of them?

Comment (katsukixyz) [reply web/db/actions/Post.ts]: We need to honor the customer's intent to delete the post, so we won't need to restore them - you can go ahead with deleting the post document. Ideally, we would mark the images that failed to delete in some way so a cron job can delete them later, but I'm not sure if we'll have the time to do that this semester.

I think for this ticket, as long you have the retry logic, we'll be good.

**Comment by uma-anand** [comment]
So I wrote the integration test for the delete post endpoint- essentially it uploads a couple of files, checks if they have been uploaded, then deletes them with the endpoint. It then checks if the MongoDB object is deleted and then if the object-related items in the bucket have been deleted. I haven't worked out how to get it working with the authentication though- I looked at all the other unit tests but none of their authentication workarounds seem to correspond with this one. The `fileUpload` test that I based a lot of this on doesn't work with authentication either. I've currently just made it skip the `post.int.test` file for now- but if you remove the authentication the test seems good.

File: web/__tests__/pages/api/post/post.int.test.ts [hunk 4]
```diff
@@ -0,0 +1,154 @@
+import { readFileSync } from "fs";
+import { createRandomPost } from "../../../../db/actions/__mocks__/Post";
+import { consts } from "../../../../utils/consts";
+import storageClient from "../../../../db/storageConnect";
+
+const noResizeData = readFileSync("./__tests__/assets/no-resize.png");
+const resizeData = readFileSync("./__tests__/assets/resize.png");
+const videoData = readFileSync("./__tests__/assets/video.mp4");
+
+jest.setTimeout(10000);
```
Comment (uma-anand) [inline web/__tests__/pages/api/post/post.int.test.ts]: The timeout seems to have caused the build error because the logs say it's a problem with `jest`, so I'll remove that.

Comment (uma-anand) [reply web/__tests__/pages/api/post/post.int.test.ts]: That doesn't seem to have fixed it, that's strange.

File: web/db/actions/Post.ts [hunk 5]
```diff
@@ -83,23 +90,46 @@ async function getResizedUploadUrl(uuid: string): Promise<string> {
...
-  const deleteObjectsCommand = new DeleteObjectsCommand({
-    Bucket: consts.storageBucket,
-    Delete: { Objects: objectsToDelete },
-  });
-  try {
-    const returned = await storageClient.send(deleteObjectsCommand);
-    return { success: true };
-  } catch (error) {
-    return { success: false };
+  for (let i = 0; i < 2; i++) {
```
Comment (katsukixyz) [inline web/db/actions/Post.ts]: Create a variable for max retries and use a while loop for readability

Comment (katsukixyz) [reply web/db/actions/Post.ts]: The retry logic should only cover what's in the try catch right now - you won't need to remap the `objectsToDelete` and `deleteObjectsCommand`

File: web/db/actions/Post.ts [hunk 6]
```diff
@@ -83,23 +90,46 @@ async function getResizedUploadUrl(uuid: string): Promise<string> {
...
+          element["Key"] &&
+            !element["DeleteMarker"] &&
+            arr.push(element["Key"]);
+        });
+        keysToDelete = arr;
+      } else {
+        return { success: true };
+      }
+    } catch (error) {
+      return { success: false };
```
Comment (katsukixyz) [inline web/db/actions/Post.ts]: throw an error on failure

File: web/db/actions/Post.ts [hunk 7]
```diff
@@ -83,23 +90,46 @@ async function getResizedUploadUrl(uuid: string): Promise<string> {
...
+  try {
+    const post = await getPost(id, false);
+    const returned = await deleteAttachments(post.attachments);
+    if (returned.success) {
+      const deletePost = await Post.deleteOne({ _id: id });
+      if (deletePost.deletedCount === 1) {
+        return { success: true };
+      }
+    }
+  } catch (e) {}
```
Comment (katsukixyz) [inline web/db/actions/Post.ts]: if `e instanceof TRPCError` throw e, otherwise throw internal server error

File: web/db/actions/Post.ts [hunk 8]
```diff
@@ -83,23 +90,46 @@ async function getResizedUploadUrl(uuid: string): Promise<string> {
...
+  return { success: false };
```
Comment (katsukixyz) [inline web/db/actions/Post.ts:132]: throw an error on failure

File: web/db/actions/Post.ts [hunk 9]
```diff
@@ -83,23 +90,46 @@ async function getResizedUploadUrl(uuid: string): Promise<string> {
...
+    try {
+      const returned = await storageClient.send(deleteObjectsCommand);
+      let arr: string[] = [];
+      if (returned["Deleted"]) {
+        returned["Deleted"].forEach((element) => {
+          element["Key"] &&
+            !element["DeleteMarker"] &&
+            arr.push(element["Key"]);
+        });
+        keysToDelete = arr;
```
Comment (katsukixyz) [inline web/db/actions/Post.ts]: This logic is a bit hard to follow - you shouldn't need to reset `keysToDelete` based on which items were deleted.

Per https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/clients/client-s3/classes/deleteobjectscommand.html:
> For each key, Amazon S3 performs a delete action and returns the result of that delete, success, or failure, in the response. **Note that if the object specified in the request is not found, Amazon S3 returns the result as deleted.**

Consider simplifying this to just check the number of deleted elements to determine whether we need to retry or not.

**Review by katsukixyz — COMMENTED** [review]
Good work on the changes. Sorry for reviewing this so late! Just a few more details to fix up. I will be working on the integration test issues tomorrow!!

File: web/db/actions/Post.ts (hunk 6 above)
Comment (katsukixyz) [reply web/db/actions/Post.ts]: You can also consider adding a `TODO` comment mentioning that we may benefit from writing a cron job to mark unsuccessful deletions to delete later

File: web/db/actions/Post.ts [hunk 10]
```diff
@@ -82,6 +90,54 @@ async function getResizedUploadUrl(uuid: string): Promise<string> {
   return `${consts.baseUrl}/api/resizedUpload/${token}`;
 }
 
+async function deleteAttachments(keysToDelete: string[]) {
+  let numTries = 1;
```
Comment (katsukixyz) [inline web/db/actions/Post.ts:94]: set equal to 0 at the beginning

File: web/db/actions/Post.ts [hunk 11]
```diff
@@ -82,6 +90,54 @@ async function getResizedUploadUrl(uuid: string): Promise<string> {
...
+      const returned = await storageClient.send(deleteObjectsCommand);
+      if (!returned["Deleted"]) {
+        return { success: true };
+      }
+    } catch (error) {
+      //TODO: Write cron job to mark unsuccessful deletions to delete later
+      throw new Error("All attachments not successfully deleted");
+    }
+    numTries++;
+  }
```
Comment (katsukixyz) [inline web/db/actions/Post.ts]: Update this loop to take the form:
```js
let count = 0;
const maxTries = 3;
while (true) {
  try {
    logic
  } catch (e) {
    if (numTries++ === maxTries) throw new Error(...)
  }
}
```
Right now we throw an error on the first try - we want to only throw an error after we have reached out maximum number of tries.

File: web/db/actions/Post.ts [hunk 12]
```diff
@@ -82,6 +90,54 @@ async function getResizedUploadUrl(uuid: string): Promise<string> {
...
+    if (e instanceof TRPCError) {
+      throw e;
+    } else {
+      throw new TRPCError({
+        message: "Internal Server Error",
+        code: "INTERNAL_SERVER_ERROR",
+      });
+    }
+  }
+  return { success: false };
```
Comment (katsukixyz) [inline web/db/actions/Post.ts]: This code is unreachable, you can delete this line

File: web/db/actions/Post.ts [hunk 13]
```diff
@@ -82,6 +90,54 @@ async function getResizedUploadUrl(uuid: string): Promise<string> {
...
+    numTries++;
+  }
+  return { success: true };
+}
+
+async function deletePost(id: ObjectId) {
+  try {
+    const post = await getPost(id, false);
+    const returned = await deleteAttachments(post.attachments);
+    if (returned.success) {
```
Comment (katsukixyz) [inline web/db/actions/Post.ts]: You can remove this since we either throw an error or it succeeds. If it errors, we enter the catch anyway.

File: web/db/actions/Post.ts [hunk 14]
```diff
@@ -82,6 +90,54 @@ async function getResizedUploadUrl(uuid: string): Promise<string> {
...
+      const deletePost = await Post.deleteOne({ _id: id });
+      if (deletePost.deletedCount === 1) {
```
Comment (katsukixyz) [inline web/db/actions/Post.ts]: Negate this case, if it's not 1 then we throw an error instead

File: web/db/actions/Post.ts [hunk 15]
```diff
@@ -131,6 +187,15 @@ async function getAllPosts() {
   return await Post.find().sort({ date: -1 });
 }
 
+async function getAttachments(oid: ObjectId) {
+  const listObjectsCommand = new ListObjectsCommand({
+    Bucket: consts.storageBucket,
+    Prefix: oid + "",
```
Comment (katsukixyz) [inline web/db/actions/Post.ts]: What's the purpose of `+ ""`? Is it to turn `oid` into a string? If so, you can try using type assertion to assert `oid as string`

File: web/server/routers/post.ts [hunk 16]
```diff
@@ -198,6 +199,22 @@ export const postRouter = router({
...
+    )
+    .mutation(async ({ input }) => {
+      const deleteSuccess = await deletePost(input.postOid);
+      if (!deleteSuccess.success) {
+        throw new TRPCError({
+          code: "INTERNAL_SERVER_ERROR",
+          message: "An unexpected error occurred.",
+        });
+      }
+      return { success: true };
```
Comment (katsukixyz) [inline web/server/routers/post.ts]: Wrap in a try catch instead of checking the `success` value. Add logic to throw the error directly if it is an instance of `TRPCError`, and internal server error otherwise

**Review by katsukixyz — COMMENTED** [review]
I think I found some things that I missed previously - my bad! I'm sorry I haven't gotten back to you on the jest stuff. If I can't get to it by tonight I'll go ahead and approve your PR and I'll make a note on working on it later.

File: web/db/actions/Post.ts (hunk 15 above)
Comment (uma-anand) [reply web/db/actions/Post.ts]: Yes, it was. It was because `oid as string` threw an error. I've changed it to `oid as unknown as string` now. Let me know if there's a way to do it without the `unknown` intermediary cast.

Comment (katsukixyz) [reply web/db/actions/Post.ts]: Alright, gotcha. Could you try `oid.toString()`?

File: web/db/actions/Post.ts [hunk 17]
```diff
@@ -82,6 +90,57 @@ async function getResizedUploadUrl(uuid: string): Promise<string> {
...
+      if (!returned["Deleted"]) {
+        return { success: true };
+      }
+    } catch (error) {
+      //TODO: Write cron job to mark unsuccessful deletions to delete later
+      if (numTries++ == maxTries) {
+        throw new Error("All attachments not successfully deleted");
+      }
+    }
+    numTries++;
```
Comment (katsukixyz) [inline web/db/actions/Post.ts]: You can safely delete this line - only increment `numTries` on error

File: web/db/actions/Post.ts [hunk 18]
```diff
@@ -82,6 +90,57 @@ async function getResizedUploadUrl(uuid: string): Promise<string> {
...
+  const deleteObjectsCommand = new DeleteObjectsCommand({
+    Bucket: consts.storageBucket,
+    Delete: { Objects: objectsToDelete, Quiet: true },
+  });
+  while (true) {
+    try {
+      const returned = await storageClient.send(deleteObjectsCommand);
+      if (!returned["Deleted"]) {
+        return { success: true };
+      }
```
Comment (katsukixyz) [inline web/db/actions/Post.ts]: There's one thing I didn't catch previously - we need to check the length of the `Deleted` field to see if it matches `keysToDelete.length`. 

```js
if (returned.Deleted?.length === keysToDelete.length) {
  return { success: true };
} else {
  throw new Error("All attachments not successfully deleted.");
}
```

File: web/db/actions/Post.ts [hunk 19]
```diff
@@ -82,6 +90,57 @@ async function getResizedUploadUrl(uuid: string): Promise<string> {
...
+    } catch (error) {
+      //TODO: Write cron job to mark unsuccessful deletions to delete later
+      if (numTries++ == maxTries) {
+        throw new Error("All attachments not successfully deleted");
```
Comment (katsukixyz) [inline web/db/actions/Post.ts]: You can reduce this to `throw e`

File: web/db/actions/Post.ts [hunk 20]
```diff
@@ -190,7 +193,7 @@ async function getAllPosts() {
 async function getAttachments(oid: ObjectId) {
   const listObjectsCommand = new ListObjectsCommand({
     Bucket: consts.storageBucket,
-    Prefix: oid + "",
+    Prefix: oid as unknown as string,
```
Comment (katsukixyz) [inline web/db/actions/Post.ts]: See comment thread from earlier

File: web/db/actions/Post.ts (hunk 18 above)
Comment (uma-anand) [reply web/db/actions/Post.ts]: The `Deleted` field only returns the elements that haven't been deleted. Also, previously in a previous comment, you had mentioned to go ahead with post deletion even if all the attachments hadn't been deleted (potentially writing a cron job to go through those). The way we've modified it now in around the last two commits, if it's a partial deletion we throw an error for the whole thing and do not proceed with post deletion. Is that okay?

Comment (katsukixyz) [reply web/db/actions/Post.ts]: Yes, you're right. Completely forgot about quiet mode. Let's revert back to the default mode so we get the results of every key. I'm thinking that we should still throw an error from the database action but just add catch when calling `deleteAttachments`:

`deleteAttachments` should stay like this:
```js
try {
  const returned = await storageClient.send(deleteObjectsCommand);
  if (returned.Deleted?.length === keysToDelete.length) {
    return { success: true };
  } else {
    throw new Error("All attachments not successfully deleted.");
  }
} catch (e) {
  //TODO: Write cron job to mark unsuccessful deletions to delete later
  if (numTries++ == maxTries) {
    throw e;
  }
}
```

`deletePost`:
```js
const post = await getPost(id, false);
>>> const returned = deleteAttachments(post.attachments).catch((e) => Sentry.captureException(e));
const deletePost = await Post.deleteOne({ _id: id });
if (deletePost.deletedCount !== 1) {
  throw new Error("Post document deletion failed");
}
```

Comment (katsukixyz) [reply web/db/actions/Post.ts]: I removed the try-catch in `deletePost` since if there is an error here (that's not an error from `deleteAttachments`), it will get caught properly by `post.delete` the way you've handled it.

Comment (uma-anand) [reply web/db/actions/Post.ts]: So the way the `DeleteObjectsCommand` works with quiet mode false, it will return a list of every object under deleted whether successful or not. So `returned.Deleted?.length === keysToDelete.length` will always be true unless the `DeleteObjectsCommand` throws an error which it does not do for partial deletions or even no deletions and just for server errors etc. So to make sure that all attachments are deleted, probably the best way would be to use quiet mode and check if `returned["Deleted"]` exists as it's done right now. At least that's how I've seen the `DeleteObjectsCommand` work when I was figuring it out and it seems to match up with the docs at https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/clients/client-s3/classes/deleteobjectscommand.html .

Comment (katsukixyz) [reply web/db/actions/Post.ts]: My understanding was that all the successfully deleted items go into `Deleted`, the rest into `Errors` - I can get more clarifications on that

Comment (katsukixyz) [reply web/db/actions/Post.ts]: See this: https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/Package/-aws-sdk-client-s3/Interface/DeleteObjectsCommandOutput/
image

Comment (katsukixyz) [reply web/db/actions/Post.ts]: Seems like `Deleted` can contain successfully deleted objects as well as objects that weren't found. We're not concerned about whether an object can be found or not - we want to pay attention to errors like _Access Denied_, _Forbidden_, S3 authorization issues, stuff like that.

File: web/yarn.lock
Comment (katsukixyz) [inline web/yarn.lock:1]: Keep this file unstaged and removed from the branch since we have no corresponding `package.json` changes

File: web/db/actions/Post.ts [hunk 21]
```diff
@@ -82,6 +90,45 @@ async function getResizedUploadUrl(uuid: string): Promise<string> {
...
 
+async function deleteAttachments(keysToDelete: string[]) {
+  let numTries = 1;
+  const maxTries = 3;
+  const objectsToDelete = keysToDelete.map((keyToDelete) => ({
+    Key: keyToDelete,
+  }));
+  const deleteObjectsCommand = new DeleteObjectsCommand({
+    Bucket: consts.storageBucket,
+    Delete: { Objects: objectsToDelete, Quiet: true },
```
Comment (katsukixyz) [inline web/db/actions/Post.ts]: Update `Quiet: false`

**Review by katsukixyz — APPROVED** [review]
Sorry this approval was slow - good work!
